### PR TITLE
Pruthvi/bn inference

### PR DESCRIPTION
**Pull request by pruthviIntel**
- batchnorm inference support in bn ctor
- layout propagation
- mkldnn emitter code for bn inference

### CODE REVIEW

File: src/ngraph/ops/batch_norm.cpp [hunk 1]
```diff
@@ -62,12 +63,90 @@ ngraph::op::BatchNorm::BatchNorm(double eps,
...
+    : RequiresTensorViewArgs("BatchNorm", {gamma, beta, input, mean, variance})
+    , m_bn_input_shape(input->get_shape())
+    , m_bn_variance_shape(variance->get_shape())
+    , m_bn_mean_shape(mean->get_shape())
+    , m_epsilon(eps)
+    , m_training(false)
+{
+    if (m_bn_input_shape.size() < 2)
+    {
+        throw ngraph_error("input tensor to batchnorm much have tensor of atleast rank 2");
```
Comment (aprocter) [inline src/ngraph/ops/batch_norm.cpp]: `much` -> `must`
`atleast` -> `at least`

As long as we're here, I noticed that the other `BatchNorm` constructor does not check the input's element type. Can we remedy that?

File: src/ngraph/ops/batch_norm.cpp [hunk 2]
```diff
@@ -62,12 +63,90 @@ ngraph::op::BatchNorm::BatchNorm(double eps,
...
+    }
+
+    if (m_bn_input_shape[1] == 0)
+    {
+        throw ngraph_error(
+            "input tensor must have atleast one channel axis for batch normalization");
```
Comment (aprocter) [inline src/ngraph/ops/batch_norm.cpp]: `atleast` -> `at least`

File: src/ngraph/ops/batch_norm.cpp [hunk 3]
```diff
@@ -62,12 +63,90 @@ ngraph::op::BatchNorm::BatchNorm(double eps,
...
+    }
+
+    if (variance->get_shape()[0] != mean->get_shape()[0])
+    {
+        throw ngraph_error("mean and variance should have same size");
+    }
+
+    if (gamma->get_shape().size() != beta->get_shape().size())
+    {
+        throw ngraph_error("gamma and beta rank does not match");
```
Comment (aprocter) [inline src/ngraph/ops/batch_norm.cpp]: This check is redundant, given the loop above.

File: src/ngraph/ops/batch_norm.cpp [hunk 4]
```diff
@@ -62,12 +63,90 @@ ngraph::op::BatchNorm::BatchNorm(double eps,
...
+    }
+
+    add_output(input->get_element_type(), m_bn_input_shape);
+}
+
```
Comment (aprocter) [inline src/ngraph/ops/batch_norm.cpp]: ^^^ I don't think we're checking here if `m_bn_input_shape[1]` matches the shapes of `gamma`, `beta`, `mean`, and `variance`.

File: src/ngraph/ops/batch_norm.cpp [hunk 5]
```diff
@@ -62,12 +63,90 @@ ngraph::op::BatchNorm::BatchNorm(double eps,
...
+        return std::make_shared<BatchNorm>(m_epsilon,
+                                           new_args.at(0),
+                                           new_args.at(1),
+                                           new_args.at(2),
+                                           new_args.at(3),
+                                           new_args.at(4));
+    }
 }
 
 ngraph::op::BatchNormBackprop::BatchNormBackprop(double eps,
```
Comment (jbobba) [inline src/ngraph/ops/batch_norm.cpp]: Can you throw an ngraph_error if generate_adjoints is called on an inference batchnorm

File: src/ngraph/runtime/cpu/mkldnn_emitter.cpp [hunk 6]
```diff
@@ -572,29 +573,52 @@ size_t MKLDNNEmitter::build_batchnorm_forward(const mkldnn::memory::desc& input_
                                               const mkldnn::memory::desc& result_desc,
                                               const mkldnn::memory::desc& mean_desc,
                                               const mkldnn::memory::desc& variance_desc,
-                                              const double eps)
+                                              const double eps,
+                                              const ngraph::op::BatchNorm* bn_node)
```
Comment (jbobba) [inline src/ngraph/runtime/cpu/mkldnn_emitter.cpp]: Dont pass the node. Flag should be sufficient right?

File: src/ngraph/ops/batch_norm.cpp (hunk 2 above)
Comment (Krovatkin) [reply src/ngraph/ops/batch_norm.cpp]: @aprocter we are saving precious bytes here!

**Comment by Krovatkin** [comment]
@pruthviIntel are there no changes to serializer?

**Comment by DawnStone** [comment]
TESTNOW

**Comment by pruthviIntel** [comment]
@jbobba @aprocter addressed your comments.
